Experiment in lock-free module invalidation

hubertp opened the pull request:
### Pull Request Description

1. Experimenting with invalidating modules' indexes without requiring full write-context locks. That should significantly improve the execution.
2. Improving performance by making background job executor run in a larger threadpool than 1.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

hubertp committed:
Experiment in lock-free module invalidation

Experimenting with invalidating modules' indexes without requiring full
write-context locks. That should significantly improve the execution.

hubertp committed:
Increae pool for background jobs

Deserialization of suggestions sounds like a fully parallelizable job
that be done independently for individual libraries. In fact it looks
like all jobs have sufficient synchronization primitives to avoid
problems and we can easily increase the pool.

4e6 commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/AnalyzeModuleInScopeJob.scala:
```diff
@@ -60,8 +61,13 @@ final class AnalyzeModuleInScopeJob(
       )
-      sendModuleUpdate(notification)
-      module.setIndexed(true)
+      if (!module.indexed()) {
```
Now I see that because it is effectively the only place that marks the module as indexed, just synchronizing around the indexed flag should be enough :+1:

JaroslavTulach replied:
> [Line 44](https://github.com/enso-org/enso/pull/9639#discussion_r1556902321) is setting an internal module field and leaving synchronized section...

Now we are changing the field again. Meanwhile the state wasn't `synchronized`. What makes you believe the result of such operation is going to be consistent? Btw.

What is the _invariant to be consistent_ with?

hubertp committed:
executors with non-empty queue

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/ThreadExecutors.java:
```diff
@@ -39,6 +39,22 @@ ExecutorService newCachedThreadPool(String name, boolean systemThread, int min,
+  ExecutorService newCachedThreadPool(
+      String name, boolean systemThread, int min, int max, int maxQueueSize) {
+    assert min >= 0;
+    assert max <= Integer.MAX_VALUE;
```
A bit useless check for an `int`, don't you think?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -495,15 +503,37 @@ public void renameProject(String newName) {
-  public void setIndexed(boolean indexed) {
-    isIndexed = indexed;
+  /** Marks the module as not indexed, should be set at the beggining of the indexing operation. */
+  public void indexing() {
```
Wrong name of the method. It shouldn't be `indexing` when it _Marks the module as not indexed_.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -495,15 +503,37 @@ public void renameProject(String newName) {
    */
   public boolean isIndexed() {
-    return isIndexed;
+    synchronized (indexingLock) {
```
I don't see a point in having `synchronized (indexingLock)`. There are two reasons to use synchronization:
- ensure consistency of (complex) [monitors](http://wiki.apidesign.org/wiki/Monitor)
- make sure changes are propagated to other threads

@hubertp, why have you decided to use `synchronized (indexingLock)` here?

As single `indexState` filled with an immutable `enum` is hardly callable a _complex monitor_ and if you just want the field value to be visible, `volatile` would work as well and it is (probably) cheaper.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala:
```diff
@@ -50,12 +50,13 @@ final class JobExecutionEngine(
       "prioritized-job-pool",
       2,
-      Integer.MAX_VALUE,
+      4,
```
What does this change in `prioritized-job-pool` and `background-job-pool` gives us?

hubertp replied:
I don't like the idea of the pool growing infinitely. Behind the scenes _cached_ threadpools  were assuming _infinite_ number of threads and a zero size blocking queue. That, I believe, wasn't correct as it would easily hide problems in execution without making it visible. I believe it's better to have a fixed maximal size of threadpool and queue of tasks that can grow up to some point.
If the queue of tasks gets full, I picked some reasonable constant I hope, then it will apply the default policy (that is rejecting tasks). At least that way we will know that something is wrong. Either way at that point it is likely a bigger, non-recoverable, problem happened.

JaroslavTulach replied:
> don't like the idea of the pool growing infinitely.

OK to adding _upper limit_ to _unlimited pools_.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala:
```diff
@@ -50,12 +50,13 @@ final class JobExecutionEngine(
 
   private val backgroundJobExecutor: ExecutorService =
-    context.newFixedThreadPool(1, "background-job-pool", false)
+    context.newCachedThreadPool("background-job-pool", 1, 4, 20, false)
```
Using more CPUs is probably desirable, but it also opens a new space for [race conditions](http://wiki.apidesign.org/wiki/Race_condition). I'd like to know the benefits, before we go that route.

hubertp replied:
Deserialization jobs are quite expensive in terms of execution time and there is no need to execute them sequentially. 
This is particularly important when modules' invalidation is triggered.

JaroslavTulach replied:
> Deserialization jobs are quite expensive in terms of execution time

Are they? We are supposed to have _lazy deserialization_ - in such case most of the deserialization is happening later, during code execution, anyway. However I can see that `SuggestionsCache` is still relying on old `java.io.ObjectInputStream` mechanism - probably because we are not benchmarking `SuggestionsCache` in our [startup benchmarks](https://enso-org.github.io/engine-benchmark-results/stdlib-benchs.html#org_enso_benchmarks_generated_Startup_hello_world_startup).

> This is particularly important when modules' invalidation is triggered.

Maybe you want to share some profiling output to demonstrate how important this is.

Btw. `background-job-pool` is probably used for many other tasks, than deserialization - are they all ready for parallel processing?

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/AnalyzeModuleInScopeJob.scala:
```diff
@@ -41,6 +41,7 @@ final class AnalyzeModuleInScopeJob(
     ctx: RuntimeContext
   ): Unit = {
     if (!module.isIndexed && module.getSource != null) {
+      module.indexing()
```
Here we are setting an internal `module` field and leaving `synchronized` section...

JaroslavTulach reviewed (changes requested):
Please find some _smaller_ issues spread as inline comments. The bigger concern is expressed here.

Thanks for bringing this `isIndex()` and `setIndexed()` flag to my attention. I never knew what it is good for. Now it looks like being related to _suggestion database_. I am scared whenever I see synchronization being spread across multiple files. I am horrified seeing synchronization being spread across multiple projects (`Module` is in `runtime` while the main logic is in `runtime-instrument-common`).

We desperately need some _encapsulation_.

It is probably acceptable (in current state of affairs) for `Module` to hold some internal state needed by _suggestion DB_ and let `runtime-instrument-common` module invalidate it and schedule its recomputation. However it is not acceptable for a different module to  poke around the `Module` flag and change its value randomly. Making such field `synchronized` doesn't improve the situation at all.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -57,7 +57,15 @@ public final class Module implements EnsoObject {
+    Indexed
+  }
+
+  private IndexState indexState;
```
Let me assume the point of this `IndexState` is to have a _tri-state_ status which can be invalidated, scheduled for recomputation and consumed, when it is available. I believe such a concept is best expressed by a [Future](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html). If we change the code in `Module` to just:
```java
private volatile Future<Object> index;

public final void indexing(Future<Object> newIndex) {
  this.index = newIndex; 
}

public final Future<Object> index() {
  return this.index;
}
```
I'll consider it to be _encapsulated enough_ (for now). @hubertp, is such a _future based interface_ enough for your purposes?

hubertp replied:
I don't believe that encapsulated the whole logic as it wouldn't express a situation when in-progress indexing would be marked as dirty and require re-indexing. I can be obviously wrong but there was a reason why logic was added to `Analyze*Job` to deal with this scenario without locking everything, as it was.

JaroslavTulach replied:
I continue to believe `Future` is the best concept to represent the desired behavior.

> it wouldn't express a situation when in-progress indexing would be marked as dirty and require re-indexing

No, (cancellable) `Future` represents such concepts neatly, I believe.

hubertp committed:
Fix broken logic

hubertp committed:
PR Review: Extract indexing logic

Extracted logic used for indexing modules to a separate map available
through `RuntimeContext`. That wasy should avoid mixing in LS logic into
runtime `Module`.
Adjusted threadpools to better reflect the needs.

hubertp commented:
> Please find some _smaller_ issues spread as inline comments. The bigger concern is expressed here.

This was a useful feedback although I don't believe I have made it worse than it used to be. Previously we would use a rather broad `writeCompilationLock` with little consideration on how that affects scheduling of tasks. 
Either way, I extracted the logic out of the runtime `Module`, as I agree that this is not an ideal place for _indexing_ flag. Instead, similarly to `PendingEdits`, it is now carried around in `RuntimeContext`. Let me know what you think.

JaroslavTulach commented:
> I extracted the logic out of the runtime `Module`, as I agree that this is not an ideal place for _indexing_ flag. Instead, similarly to `PendingEdits`, it is now carried around in `RuntimeContext`.

From a code perspective, such a change doesn't have real impact on the behavior. Whatever flaws were present so far, are still present. The only real impact is _organizational one_: This PR moves the code away from `engine/runtime` project - the changes in `engine/runtime` project are non-controversial now.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ModuleIndexing.scala:
```diff
@@ -0,0 +1,48 @@
+    val NotIndexed, NeedsIndexing, Indexed = Value
+  }
+
+  private val modules: ConcurrentMap[Module, IndexState.Value] =
```
The code is now rewritten into a new `ModuleIndexing` Scala class that uses `ConcurrentHashMap`. Good encapsulation, but I still miss the answer to the [fundamental question](https://github.com/enso-org/enso/pull/9639#discussion_r1556904725): What is the **invariant** to be consistent with?

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/AnalyzeModuleInScopeJob.scala:
```diff
@@ -60,8 +61,13 @@ final class AnalyzeModuleInScopeJob(
       )
-      sendModuleUpdate(notification)
-      module.setIndexed(true)
+      if (!ctx.state.suggestions.markAsIndexed(module)) {
```
The problem I see [continues to be the same](https://github.com/enso-org/enso/pull/9639#discussion_r1556904725). Line 44 deals with `state.suggestions` and line 64 does that as well. But meanwhile anything can happen. What if a new request to `markAsNotIndexed` comes in another thread?

> there was a reason why logic was added to Analyze*Job to deal with this scenario

My experience with Enso code base and its _readiness for concurrent execution_ doesn't support claims that advocate _reason_ and/or _logic_. Even if they were present, it is questionable they were used in a _sound way_. Except using the most advanced concurrent building blocks, I only see attempts to avoid answers to the basic question: _how that is supposed to work at all_?

JaroslavTulach reviewed (changes requested):
Visual inspection of `Analyze*Job` line 44 and line 64 still hints there is a [possibility of race condition](https://github.com/enso-org/enso/pull/9639#discussion_r1556904725). The code might have been working mostly correctly so far (as there was just a single threaded background pool). However, as the [throughput of the  background thread pool is about to be increased from 1](https://github.com/enso-org/enso/pull/9639#discussion_r1556889025) the race conditions are likely to appear - until we justify by spelling out some _invariants_/_constraints_ [why they cannot](https://github.com/enso-org/enso/pull/9639#discussion_r1556904725)...

hubertp committed:
Eliminate race-conditions when indexing

Introducing an immutable `IndexState` record that allows us to determine
if the just computed index is really up-to-date and could be sent to the
clients.

hubertp committed:
docs

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/ModuleIndexing.java:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  public static ModuleIndexing getInstance() {
+    return new ModuleIndexing();
```
If this always creates new instance, then the method should be named `create` not `get`.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/ModuleIndexing.java:
```diff
@@ -0,0 +1,89 @@
+   */
+  public boolean markAsIndexed(Module module, IndexState state) {
+    var computed = modules.compute(module, (k, v) -> v == state ? state.toIndexed() : v);
+    return computed.from() == state;
```
I was afraid there  is a race condition between line 67 and 68. The method may return `false` even when the state was updated. But no, there is no race condition.

Isn't there however possibility of `NullPointerException`? What if the `module` isn't in the `modules` map?

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/AnalyzeModuleInScopeJob.scala:
```diff
@@ -37,31 +38,43 @@ final class AnalyzeModuleInScopeJob(
       )
-      sendModuleUpdate(notification)
-      module.setIndexed(true)
+      if (ctx.state.suggestions.markAsIndexed(module, state)) {
```
OK, this looks race condition free.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/ModuleIndexing.java:
```diff
@@ -0,0 +1,89 @@
+   * @param isIndexed true, if IR has been already indexed. False otherwise
+   * @param ir IR of a module that has been/needs to be indexed
+   */
+  public record IndexState(boolean isIndexed, IR ir, IndexState from) {
```
Will any `IndexState` ever garbage collect? I doubt when each new `IndexState` points to a chain of previous ones.

The `from` field is only needed until it is applied into the `modules` map. Then it can be cleared...

hubertp replied:
Good catch, fixed.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/ModuleIndexing.java:
```diff
@@ -0,0 +1,89 @@
+  /**
+   * @return true, if module has been isIndexed. False otherwise.
+   */
+  public boolean isIndexed(Module module) {
```
API wise I'd prefer `IndexState find(Module module)` and then operations on the `IndexState` itself...

JaroslavTulach reviewed (changes requested):
the trick with _previous index state_ is cute, but I am afraid there is a GC problem. We need to be able to GC no longer needed `IndexState` instances.

hubertp committed:
Avoid GC problems

hubertp committed:
naming is hard

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/ModuleIndexing.java:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  /**
+   * @return true, if module has been isIndexed. False otherwise.
```
The return value isn't `true` anymore.

hubertp committed:
Merge branch 'develop' into wip/hubert/lock-free-invalidation

hubertp committed:
Update doc
